addons/stock/models/stock_quant.py

```diff
@@ -292,6 +288,21 @@ def _update_reserved_quantity(self, product_id, location_id, quantity, lot_id=No
                 break
         return reserved_quants
 
+    @api.model
+    def _unlink_zero_quants(self):
+        """ _update_available_quantity may leave quants with no
+        quantity and no reserved_quantity. It used to directly unlink
+        these zero quants but this proved to hurt the performance as
+        this method is often called in batch and each unlink invalidate
+        the cache. We defer the calls to unlink in this method.
+        """
+        precision_digits = max(6, self.env.ref('product.decimal_product_uom').digits * 2)
+        # Use a select instead of ORM search for UoM robustness.
+        query = """SELECT id FROM stock_quant WHERE round(quantity::numeric, %d) = 0 AND round(reserved_quantity::numeric, %d) = 0;""" % (precision_digits, precision_digits)
+        self.env.cr.execute(query)
+        quant_ids = self.env['stock.quant'].browse([quant['id'] for quant in self.env.cr.dictfetchall()])
```

Comment:
query = """SELECT id FROM stock_quant WHERE round(quantity::numeric, %d) = 0 AND round(reserved_quantity::numeric, %d) = 0;"""
        params = (precision_digits, precision_digits)
        self.env.cr.execute(query, params )

        quants_ids = self.env['stock.quant'].browse( i for [i] in self.env.cr.fetchall() )